### DISCUSSION TITLE
Questing regarding `sum_of_costs` for PELT

### DISCUSSION THREAD

**Discussion by YanniPapandreou** [deepcharles/ruptures#203] [Q&A] [answered]
Hi,

I am new to change point detection and I have been greatly enjoying trying out your package, so thanks for creating this!

I have a quick question regarding the method `sum_of_costs` when applied to the PELT search method. I include some code below for reference:

```
algo = rpt.Pelt(custom_cost=c).fit(signal)
my_bkps = algo.predict(pen=3)
my_cost = algo.cost.sum_of_costs(my_bkps)
```
Here I have used PELT to find changepoints in a signal using a custom cost. I then use `sum_of_costs` to get the sum of costs for this segmentation. My question is the following: does `sum_of_costs` return the costs of the segments **without** the penalty term, or is this included as well? Sorry if this question is trivial, but I wasn't sure from the source code.

Best wishes,
Yanni

**Comment by deepcharles** [comment, marked as answer]
Thanks for using `ruptures`, glad to hear it is useful.

> does sum_of_costs return the costs of the segments without the penalty term

`sum_of_costs` only returns the sum of costs **without** the penalty term. This is a bit hidden in the code (it is here if you are curious: https://github.com/deepcharles/ruptures/blob/b84df2c83a33a5947b52458e1a2555701f18150a/src/ruptures/base.py#L53)

Cheers

Charles

**Reply by YanniPapandreou** [reply]
Thanks for clarifying!

Cheers,
Yanni

**Comment by frankcharltonsonos** [comment]
@deepcharles - I'm a little confused by the explanation here. I seem to get different sum_of_costs responses after adjusting the penalty for the same dataset. It appears that choice of penalty _does_ affect the sum_of_costs return. I'm hoping this means it genuinely calculates the value, based on the penalty.

```
import numpy as np
import matplotlib.pylab as plt
import ruptures as rpt

# creation of data
n, dim = 52, 3
n_bkps, sigma = 3, 1
signal, bkps = rpt.pw_constant(n, dim, n_bkps, noise_std=sigma)

# change point detection
model = "l2"  # "l2", "rbf"
algo = rpt.Pelt(model=model, jump=1).fit(signal)

my_bkps_0 = algo.predict(pen=0)
algo.cost.sum_of_costs(my_bkps_0)

my_bkps_1 = algo.predict(pen=1)
algo.cost.sum_of_costs(my_bkps_1)

my_bkps_10 = algo.predict(pen=10)
algo.cost.sum_of_costs(my_bkps_10)

my_bkps_20 = algo.predict(pen=20)
algo.cost.sum_of_costs(my_bkps_20)
```

Each of these 4 penalty / sum_of_costs runs return a different value.

**Comment by deepcharles** [comment]
Hi,

The method `.sum_of_costs(bkps)` returns the sum of costs for the given set of change-points `bkps`, i.e.  `c(0..t_1) + c(t_1..t_2) + c(t_2..t_3) + ...` where `t_1`, `t_2`,... are the change-point indexes contained in `bkps`. 

`sum_of_costs` only depends on the cost function and the argument `bkps`, i.e. the set of changes (and the signal). It is independent of the value of `pen`.

In your example, `algo.cost.sum_of_costs(my_bkps)` changes value because the argument (`my_bkps`) changes too. 

Hope this helps,

Charles

**Comment by frankcharltonsonos** [comment]
Thank you for your response -  I'm hoping this means that they are still related, even if not directly passed into the function.

Each penalty run produces a different set of `bkps`..._because_ the penalty is different. By passing an array of different penalties, I'd think that I'd be able to find the elbow for sum_of_costs associated with the 'ideal' number of changepoints. This just happens to also show me the _penalty range_ associated with that 'ideal' number of changepoints, thereby showing me what penalty to use.

This seems to be working. I'm just hoping to confirm that this is working properly, or if using `sum_of_costs` with PELT (which requires a penalty value) is returning something unusable.

**Comment by frankcharltonsonos** [comment]
Realizing now that a plot of penalty values by number of changepoints is more helpful here. Sum of costs is not as helpful.

**Comment by deepcharles** [comment]
Here is an example where the sum of costs is plotted to find the number of changes : [https://centre-borelli.github.io/ruptures-docs/examples/music-segmentation/](https://centre-borelli.github.io/ruptures-docs/examples/music-segmentation/).
